data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/client/CloudWatchLogsService.java

```diff
@@ -0,0 +1,118 @@
+    private final Executor sinkThreadManager;
+    private final CloudWatchLogsDispatcher dispatcher;
+
+    public CloudWatchLogsService(final Buffer buffer,
+                                 final CloudWatchLogsLimits cloudWatchLogsLimits,
+                                 final CloudWatchLogsDispatcher dispatcher, BlockingQueue<ThreadTaskEvents> blockingQueue) {
+
+        this.buffer = buffer;
+        this.dispatcher = dispatcher;
+        this.bufferedEventHandles = new ArrayList<>();
+        this.cloudWatchLogsLimits = cloudWatchLogsLimits;
+        this.taskQueue = blockingQueue;
+
+        bufferLock = new ReentrantLock();
+        sinkStopWatch = new SinkStopWatch();
+        sinkThreadManager = newCachedThreadPool();
+    }
+
+    /**
+     * Function handles the packaging of events into log events before sending a bulk request to CloudWatchLogs.
+     * Implements simple conditional buffer. (Sends once batch size, request size in bytes, or time limit is reached)
+     * @param logs - Collection of Record events which hold log data.
+     */
+    public void processLogEvents(final Collection<Record<Event>> logs) {
+        sinkStopWatch.startIfNotRunning();
+        for (Record<Event> log: logs) {
+            int logLength = log.getData().toJsonString().length();
+
+            if (cloudWatchLogsLimits.isGreaterThanMaxEventSize(logLength)) {
+                LOG.warn("Event blocked due to Max Size restriction! {Event Size: {} bytes}", (logLength + CloudWatchLogsLimits.APPROXIMATE_LOG_EVENT_OVERHEAD_SIZE));
+                continue;
+            }
+
+            long time = sinkStopWatch.getStopWatchTimeSeconds();
+
+            bufferLock.lock();
+
+            int bufferSize = buffer.getBufferSize();
+            int bufferEventCount = buffer.getEventCount();
+            int bufferEventCountWithEvent = bufferEventCount + 1;
+            int bufferSizeWithAddedEvent = bufferSize + logLength;
+
+            if ((cloudWatchLogsLimits.isGreaterThanLimitReached(time, bufferSizeWithAddedEvent, bufferEventCountWithEvent) && (bufferEventCount > 0))) {
+                stageLogEvents();
+                addToBuffer(log);
+            } else if (cloudWatchLogsLimits.isEqualToLimitReached(bufferSizeWithAddedEvent, bufferEventCountWithEvent)) {
+                addToBuffer(log);
+                stageLogEvents();
+            } else {
+                addToBuffer(log);
+            }
+
+            bufferLock.unlock();
+        }
+    }
+
+    private void stageLogEvents() {
+        sinkStopWatch.stopAndResetStopWatch();
+
+        ArrayList<byte[]> eventMessageCloneList = new ArrayList<>();
+        cloneLists(buffer.getBufferedData(), eventMessageCloneList);
+
+        ThreadTaskEvents dataToPush = new ThreadTaskEvents(eventMessageCloneList, bufferedEventHandles);
+        taskQueue.add(dataToPush);
+
+        buffer.clearBuffer();
+        bufferedEventHandles = new ArrayList<>();
+
+        sinkThreadManager.execute(dispatcher);
```

Comment:
Why are we dispatching the global dispatcher? Don't we want to create a new `CloudWatchLogsDispatcher` (could also be named `CloudWatchLogsUploader`) for each batch that we flush?

Reply:
That is true, using a global object seems like it holds too many responsibilities. I will refactor this segment to create a brand new dispatcher for each task. Thank you!